From ebc1b9789866cbf52781468c7392d9fc5d1297bc Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Fri, 10 Oct 2014 21:03:12 -0400
Subject: [block] blk-mq: fix WARNING "percpu_ref_kill() called more than once!"

Message-id: <1412975015-5370-12-git-send-email-jmoyer@redhat.com>
Patchwork-id: 97466
O-Subject: [RHEL7 PATCH 11/34] blk-mq: fix WARNING "percpu_ref_kill() called more than once!"
Bugzilla: 1146660
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Mike Snitzer <snitzer@redhat.com>

This is a backport of the following commit.  This resolves bug
1146660.

  commit dd840087086f3b93ac20f7472b4fca59aff7b79f
  Author: Ming Lei <ming.lei@canonical.com>
  Date:   Fri Aug 15 23:16:32 2014 +0800

    blk-mq: fix WARNING "percpu_ref_kill() called more than once!"

    Before doing queue release, the queue has been freezed already
    by blk_cleanup_queue(), so needn't to freeze queue for deleting
    tag set.

    This patch fixes the WARNING of "percpu_ref_kill() called more than once!"
    which is triggered during unloading block driver.

    Cc: Tejun Heo <tj@kernel.org>
    Signed-off-by: Ming Lei <ming.lei@canonical.com>
    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-mq.c b/block/blk-mq.c
index 22682fb..1db9565 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -1722,14 +1722,10 @@ static void blk_mq_del_queue_tag_set(struct request_queue *q)
 {
  struct blk_mq_tag_set *set = q->tag_set;
 
- blk_mq_freeze_queue(q);
-
  mutex_lock(&set->tag_list_lock);
  list_del_init(&q->tag_set_list);
  blk_mq_update_tag_set_depth(set);
  mutex_unlock(&set->tag_list_lock);
-
- blk_mq_unfreeze_queue(q);
 }
 
 static void blk_mq_add_queue_tag_set(struct blk_mq_tag_set *set,
-- 
1.7.1